File: server/pieces/KingPiece.cpp
#include "KingPiece.h"
#include "../Game.h"

std::list<Position> KingPiece::getValidMoves()
{
    Piece* piece;
    std::list<Position> valid;
    for(int x = -1; x <= 1; x++)
    {
        for(int y = -1; y <= 1; y++)
        {
            if(x==0 && y==0) continue;
            try
            {
                piece = _game->getCell( Position(_myPos.x+x, _myPos.y+y) ); 
                if( piece == NULL)
                    valid.push_front( Position(_myPos.x+x, _myPos.y+y) );
            }catch(std::out_of_range &e) { continue; }
        }
    }

    if(!_hasMoved)
    {
        std::list<RookPiece>& rooks = _game->getRook(_white);
        int howManyMoved = 0;
        for(RookPiece& p : rooks)
        {
            if(!p.hasMoved())
            {
                Position& rPos = p.getPos();
                int dx =_myPos.x - rPos.x < 0 ? -1 : 1;
                int x = _myPos.x + dx;
                int y = _myPos.y;

                while(x!=rPos.x)
                {
                    if(_game->getCell( Position(x,y) ) != nullptr)
                        break;
                }
                if(x==rPos.x)
                {
                    valid.push_front( Position(x+dx*2,y) );
                }
            }
        }
    }
/*
    if(!_hasMoved)
    {
        std::array<RookPiece,2>& rooks = _game->getRook(_white);
        if(! rooks[0].hasMoved() )
        {
            int x = -1;
            int y = 0;
            while(_myPos.x + x != rooks[0].getPos().x)
            {
                if(_game->getCell( Position(_myPos.x+x, _myPos.y+y)) != nullptr) break;
            }
            if(_myPos.x + x == rooks[0].getPos().x)
            {
                
            }
        }
    }*/
    return valid;
}

File: client/js/Pieces/KingPiece.js
class KingPiece extends Piece {
    constructor(id,white,pos,g,forward) {
        super(id,white,pos,g,forward);
        this.visual = new KingPieceVisual(this);

        this.hasMoved = false;
    }

    getValidMoves() {
        let valid = [];

        for( let x = -1; x <= 1; x++ ) {
            for( let y = -1; y <= 1; y++) {
                if(x == 0 && y == 0) continue;
                try {
                    pieceInArea = this.pushIfAvailable(valid, new Position(
                        this.pos.x + tempX, this.pos.y + tempY), undefined);
                
                } catch( err ) { continue; }
            }
        }

        return valid;
    }

    getValidCastling() {
        if(this.hasMoved) return [];

        let valid = [];
        rooks = this.game.getRook(this.white);

        for(let rook in rooks) {
            if(rook.hasMoved()) continue;

            let rPos = rook.pos;
            let dx = this.pos.x - rPos.x < 0 ? 1 : -1;
            let x = this.pos.x + dx;
            let y = this.pos.y;

            while(x!=rPos.x) {
                if(this.game.getCell( new Position(x,y) ) != null ) {
                    break;
                }

                x+=dx;
            }

            if(x==rPos.x) {
                valid.push( new Position(this.pos.x + dx*2, y));
            }
        }

        return valid;
    }

    validateCastling(dest) {
        for(let pos in this.getValidCastling() ) {
            if(pos.x == dest.x && pos.y == dest.y) {
                return true;
            }
        }
        return false;
    }

    getCastlingRook(dest) {
        if(this.validateCastling(dest)==false) return null;
        
        //FIXME (Same FIXME In c++)
        const MAX_X = 7;
        const MIN_X = 0;
        if(dest.x == this.pos + 2) {
            return this.game.getCell( new Position(MAX_X ,this.pos.y));
        }
        else {
            return this.game.getCell( new Position(MIN_X, this,pos.y));
        }
    }

    setPos(pos) {
        super.setPos(pos);
        this.hasMoved = true;
    }

    hasMoved() { return this.hasMoved; }

    update(deltaTime) {
        this.visual.update(deltaTime);
    }
}